File: Kirigami/scripts/plot_k6.py
#!/usr/bin/env python3
"""K6 figure: inward split edges at t = 0, and the certified Theta_max distributions.

Run as:  arch -arm64 /usr/local/bin/python3 code/scripts/plot_k6.py results/kill/k6/k6.csv \
             -o results/kill/k6/k6_zero_plus.png
"""
import argparse
import csv
import sys

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np

FAMS = ["sigma_mc", "sigma_def"]
COL = {"sigma_mc": "#3b6ea5", "sigma_def": "#c2622d"}


def load(path):
    rows = {f: [] for f in FAMS}
    with open(path) as f:
        for r in csv.DictReader(f):
            if r["sigma"] in rows:
                rows[r["sigma"]].append(r)
    return rows


def col(rs, key, cast=float):
    return np.array([cast(r[key]) for r in rs]) if rs else np.array([])


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("csv")
    ap.add_argument("-o", "--out", default="k6.png")
    a = ap.parse_args()
    rows = load(a.csv)
    if not any(rows.values()):
        sys.exit("no rows")

    fig, ax = plt.subplots(1, 3, figsize=(15, 4.4))

    # (a) inward split edges at t = 0, as a fraction of |E_split|
    for f in FAMS:
        v = col(rows[f], "inward_frac0")
        if v.size:
            ax[0].hist(v, bins=np.linspace(0, 1, 26), alpha=0.6, label=f, color=COL[f])
    ax[0].set_xlabel(r"inward split edges $q_e\leq 0$ / $|E_{split}|$ at $t=0$")
    ax[0].set_ylabel("graphs")
    ax[0].set_title("(a) why K5 saw $\\Theta_{max}=0$")
    ax[0].legend()

    # (b) absolute counts vs |E_split|
    for f in FAMS:
        x, y = col(rows[f], "n_split"), col(rows[f], "inward0")
        if x.size:
            ax[1].scatter(x, y, s=12, alpha=0.6, label=f, color=COL[f])
    lim = max([col(rows[f], "n_split").max() if rows[f] else 0 for f in FAMS] + [1])
    ax[1].plot([0, lim], [0, lim], "k--", lw=0.8, label="all inward")
    ax[1].set_xlabel(r"$|E_{split}|$")
    ax[1].set_ylabel("inward split edges at $t=0$")
    ax[1].set_title("(b) count vs split-cut count")
    ax[1].legend()

    # (c) certified Theta_max after the repair, plus the refereeing bisection
    labels, data, colors = [], [], []
    for f in FAMS:
        rs = [r for r in rows[f] if int(r["feasible"])]
        for key, tag in (("eps_max", "certified"), ("theta_bisect", "bisection")):
            v = np.array([float(r[key]) for r in rs]) if rs else np.array([])
            labels.append(f"{f}\n{tag}\n(n={v.size})")
            data.append(v if v.size else np.array([0.0]))
            colors.append(COL[f])
    pos = np.arange(len(data))
    for p, d, c in zip(pos, data, colors):
        jitter = (np.random.RandomState(0).rand(d.size) - 0.5) * 0.3
        ax[2].scatter(p + jitter, d, s=14, alpha=0.7, color=c)
    ax[2].set_xticks(pos)
    ax[2].set_xticklabels(labels, fontsize=7)
    ax[2].set_ylabel(r"$\Theta_{max}$ (rad)")
    ax[2].set_title("(c) after the 0+ repair, at the $0^+$-feasible points")
    ax[2].set_ylim(-0.05, max(0.35, float(np.max([d.max() for d in data])) * 1.1))

    fig.suptitle("K6 -- 0+ repair in the Tutte auxetic null space, 200 random graphs")
    fig.tight_layout()
    fig.savefig(a.out, dpi=150)
    print("wrote", a.out)


if __name__ == "__main__":
    main()

File: Kirigami/scripts/plot_k6.jl
#!/usr/bin/env julia
# K6 figure: inward split edges at t = 0, and the certified Theta_max distributions.
# Port of plot_k6.py.
# Run: julia --project=Kirigami/scripts Kirigami/scripts/plot_k6.jl results/kill/k6/k6.csv \
#          -o results/kill/k6/k6_zero_plus.png
include(joinpath(@__DIR__, "plot_common.jl"))
using Random

const FAMS = ["sigma_mc", "sigma_def"]
const COL = Dict("sigma_mc" => "#3b6ea5", "sigma_def" => "#c2622d")

function parse_args(args)
    csvp, out = nothing, "k6.png"
    i = 1
    while i <= length(args)
        if args[i] in ("-o", "--out")
            out = args[i+1]; i += 2
        else
            csvp = args[i]; i += 1
        end
    end
    csvp === nothing && error("usage: plot_k6.jl <csv> [-o out.png]")
    return csvp, out
end

function main()
    csvp, out = parse_args(ARGS)
    byfam = Dict(f => Row[] for f in FAMS)
    for r in rows(csvp)
        haskey(byfam, r["sigma"]) && push!(byfam[r["sigma"]], r)
    end
    all(isempty, values(byfam)) && error("no rows")
    col(rs, key) = [fnum(r, key) for r in rs]

    fig = Figure(size = (1500, 440))

    # (a) inward split edges at t = 0, as a fraction of |E_split|
    ax1 = Axis(fig[1, 1]; xlabel = L"inward split edges $q_e\leq 0$ / $|E_{split}|$ at $t=0$",
               ylabel = "graphs", title = L"(a) why K5 saw $\Theta_{max}=0$")
    for f in FAMS
        v = col(byfam[f], "inward_frac0")
        isempty(v) || hist_bars!(ax1, v, collect(range(0, 1, length = 26));
                                 color = COL[f], alpha = 0.6, label = f, strokewidth = 0)
    end
    axislegend(ax1)

    # (b) absolute counts vs |E_split|
    ax2 = Axis(fig[1, 2]; xlabel = L"|E_{split}|", ylabel = L"inward split edges at $t=0$",
               title = "(b) count vs split-cut count")
    lim = 1.0
    for f in FAMS
        x, y = col(byfam[f], "n_split"), col(byfam[f], "inward0")
        isempty(x) && continue
        scatter!(ax2, x, y; markersize = 7, color = (COL[f], 0.6), label = f)
        lim = max(lim, maximum(x))
    end
    lines!(ax2, [0, lim], [0, lim]; color = :black, linestyle = :dash, linewidth = 0.8,
           label = "all inward")
    axislegend(ax2)

    # (c) certified Theta_max after the repair, plus the refereeing bisection
    labels, data, colors = String[], Vector{Float64}[], String[]
    for f in FAMS
        rs = [r for r in byfam[f] if inum(r, "feasible") != 0]
        for (key, tag) in (("eps_max", "certified"), ("theta_bisect", "bisection"))
            v = [fnum(r, key) for r in rs]
            push!(labels, "$f\n$tag\n(n=$(length(v)))")
            push!(data, isempty(v) ? [0.0] : v)
            push!(colors, COL[f])
        end
    end
    ax3 = Axis(fig[1, 3]; ylabel = L"\Theta_{max}\ \mathrm{(rad)}",
               title = L"(c) after the 0+ repair, at the $0^+$-feasible points",
               xticks = (0:length(data)-1, labels), xticklabelsize = 9)
    rng = MersenneTwister(0)
    for (p, d, c) in zip(0:length(data)-1, data, colors)
        jitter = (rand(rng, length(d)) .- 0.5) .* 0.3
        scatter!(ax3, p .+ jitter, d; markersize = 7, color = (c, 0.7))
    end
    ylims!(ax3, -0.05, max(0.35, maximum(maximum.(data)) * 1.1))

    Label(fig[0, 1:3], "K6 -- 0+ repair in the Tutte auxetic null space, 200 random graphs";
          fontsize = 14)
    savefig(fig, outpath(out))
end

main()
